### PR TITLE
[Merged by Bors] - Replace `WorldQueryGats` trait with actual gats

### PR DESCRIPTION
**Pull request by BoxyUwU**
# Objective

Replace `WorldQueryGats` trait with actual gats

## Solution

Replace `WorldQueryGats` trait with actual gats

---

## Changelog

- Replaced `WorldQueryGats` trait with actual gats

## Migration Guide

- Replace usage of `WorldQueryGats` assoc types with the actual gats on `WorldQuery` trait

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]
Looks good! The `WorldQuery` trait impls are way more legible.

Great use of deprecation. The actual changes are *very* mechanical; I'm glad to see you hit the doc strings too though.

**Comment by alice-i-cecile** [comment]
This PR is blocked on the next release of Rust, which will stabilize basic GATs.

File: crates/bevy_ecs/src/query/fetch.rs [hunk 1]
```diff
@@ -452,13 +437,23 @@ pub trait WorldQueryGats<'world> {
 pub unsafe trait ReadOnlyWorldQuery: WorldQuery<ReadOnly = Self> {}
 
 /// The `Fetch` of a [`WorldQuery`], which is used to store state for each archetype/table.
-pub type QueryFetch<'w, Q> = <Q as WorldQueryGats<'w>>::Fetch;
+#[deprecated(note = "Use `Q::Fetch<'w>` instead", since = "0.9")]
+pub type QueryFetch<'w, Q> = <Q as WorldQuery>::Fetch<'w>;
```
Comment (james7132) [inline crates/bevy_ecs/src/query/fetch.rs:441]: I don't think this is a particularly prominent user-facing, probably better to just remove this and add it to the migration notes.

Same with the other aliases.

**Review by james7132 — APPROVED** [review]
This is so much nicer. My only gripe is that `ROQueryItem<'w, Q>`  to `<<Q::ReadOnly as WorldQuery>::Item` is a readability regression.

**Comment by BoxyUwU** [comment]
I'd be fine just keeping `QueryItem` so that `<Q::ReadOnly as WorldQuery>::Item<''w>` can be shorter i dont really care what exactly we do :sweat_smile:

**Comment by james7132** [comment]
It's in these cases I wish traits supported default type aliases.

**Comment by BoxyUwU** [comment]
default type alias isnt really what we want since that would allow it to be overriden. What you'd want it something that behaves more like:
```rust
trait GetReadOnlyItem {
    type ReadOnlyItem<'w>;
    type ReadOnlyFetch<'w>;
}

impl<T: WorldQuery> GetReadOnlyItem for T {
    type ReadOnlyItem<'w> = <T::ReadOnly as WorldQuery>::Item<'w>;
    type ReadOnlyItem<'w> = <T::ReadOnly as WorldQuery>::Fetch<'w>; 
}
```

**Comment by BoxyUwU** [comment]
Actually thats tiny enough of a change that im just Gonna Do That instead of `type ROQueryItem<'a, Q>`

**Comment by BoxyUwU** [comment]
Wow that incredibly did not work and I have no idea why, guess I'm not doing that lol

**Comment by BoxyUwU** [comment]
decided against removing `QueryFetch` and `ROQueryFetch` as until my PR to remove the `Clone` bounds on `Fetch` from iter_combinations people will be required to put `where ROQueryFetch<'a, Q>: Clone` or `where QueryFetch<'a, Q>: Clone` in functions that call `iter_combinations` on generic queries. Would rather not force people to do `<<Q as WorldQuery>::ReadOnly as WorldQuery>::Fetch<'a>: Clone`

**Review by JoJoJet — APPROVED** [review]

**Comment by BoxyUwU** [comment]
looked over the changes after rebasing and they still seem correct to me

**Review by alice-i-cecile — APPROVED** [review]
Rebase LGTM.

**Review by devil-ira — APPROVED** [review]

**Pull request closed without merging**